### PR TITLE
BR-4 Test Case

### PR DESCRIPTION
**Pull request by Malankar**
This pull request includes several significant updates to the project, focusing on adding testing capabilities, improving error handling, and refining the bookmark controller. The most important changes include the addition of Vitest for testing, enhancements to the bookmark controller, and the introduction of rate limiting and error handling tests.

### Testing Enhancements:
* Added Vitest as the test runner and configured it in `vite.config.js` to include coverage reporting and specific test directories. (`vite.config.js`, [vite.config.jsR1-R33](diffhunk://#diff-58e6f63d87181b1c6a8cb6e5f1691df04aa32854456efcd52ca71c8541375d26R1-R33))
* Added test scripts and dependencies for Vitest and Supertest in `package.json`. (`package.json`, [[1]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L7-R9) [[2]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R25-R30)

### Bookmark Controller Improvements:
* Removed unnecessary error handling in `getBookmarkById` and fixed a variable typo in `createBookmark`. (`src/controllers/bookmarkController.js`, [[1]](diffhunk://#diff-1e7c6a266cf13cea39476d0b8cf37cd2b11703dab35ad0d5a1749dc521ee6a6bL15-L19) [[2]](diffhunk://#diff-1e7c6a266cf13cea39476d0b8cf37cd2b11703dab35ad0d5a1749dc521ee6a6bL30-R25)
* Added comprehensive tests for `bookmarkController` to cover all CRUD operations. (`test/controllers/bookmarkController.test.js`, [test/controllers/bookmarkController.test.jsR1-R114](diffhunk://#diff-788362b297cc977cf3088d2ce154513f062351d1096811241a9a246c7c899d2eR1-R114))

### Rate Limiting:
* Introduced rate limiting middleware in `bookmarkRoutes.js` and added tests to ensure the rate limit is enforced. (`src/routes/bookmarkRoutes.js`, [[1]](diffhunk://#diff-a3c2ebf95eb002d1e29c26cec22a8cb858de044b7565a027938a8707806b49e3R3-R5); `test/routes/bookmarkRoutes.test.js`, [[2]](diffhunk://#diff-ad0d7a4c22b71fea060056848297b22c6ec3c75a79288515e0481ae9635a5ea1R1-R27)

### Error Handling:
* Added tests for the `errorHandler` middleware to handle various error scenarios including Joi validation errors and generic errors. (`test/middleware/errorHandler.test.js`, [test/middleware/errorHandler.test.jsR1-R85](diffhunk://#diff-17c4f7dd1d13a1a463d7b5dd1b00132940edc6ac4827691e10a84d391d9311c3R1-R85))

### Database Schema Testing:
* Added tests for the `bookmark` schema to ensure the correct configuration of table columns. (`test/db/schema.test.js`, [test/db/schema.test.jsR1-R33](diffhunk://#diff-c63c5ec04dcd99e568ee6121af2b3f259a497bcfc5533ed4c55b4bc721fcdf05R1-R33))

### CODE REVIEW

File: test/routes/bookmarkRoutes.test.js
Comment (Bala-raj) [inline test/routes/bookmarkRoutes.test.js:1]: This test suite for rate limiting is a good start, but it has some significant flaws that make it unreliable and potentially misleading. Let's break down the issues and discuss improvements:

Problems:

- Timing Issues: The biggest problem is that these tests rely on actual time passing.  Your rate limiting is configured for 100 requests within 15 minutes.  Your tests are trying to send 100 requests immediately.  This is highly unlikely to reflect real-world behavior and might not even trigger the rate limiter correctly, especially if the requests are processed very quickly.  The tests are essentially just checking that 100 requests don't immediately cause a 429, which isn't a very meaningful test.

- Flaky Tests: Because of the timing issue, these tests are highly flaky.  They might pass sometimes and fail other times depending on how quickly your server processes the requests and how the rate limiter is implemented.  Flaky tests are a major problem because they erode confidence in your test suite.

- No Control over Time:  You have no way to control the passage of time within your tests.  You need a way to simulate the 15-minute window without actually waiting for 15 minutes.

- Limited Scenarios:  You're only testing the edge case of exactly 100 requests and then one more.  You should test other scenarios, such as:

    - What happens if you send 50 requests, wait for 5 minutes, and then send another 50?
    - What happens if you send 101 requests very quickly?
    - What is the behavior with different HTTP methods (GET, PUT, DELETE)? Rate limiting should ideally apply to all relevant methods.

File: test/services/bookmarkService.test.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+      const result = await bookmarkService.getAllBookmarks();
+      expect(result).toEqual([]);
+    });
+  });
+
+  describe("getBookmarkById", () => {
+    it("should return the bookmark with the given id", async () => {
+      const mockBookmark = {
+        id: "49fb6870-478f-4a6a-a839-e6e198adfb94",
```
Comment (Bala-raj) [inline test/services/bookmarkService.test.js:36]: Test Data Generation: Using hardcoded UUIDs in your tests.  This makes the tests less flexible and harder to read.  It's highly recommended to use a library like uuid to generate UUIDs within your tests.  This will make your tests more robust and less prone to errors.

File: test/services/bookmarkService.test.js [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+          where: vi.fn().mockResolvedValue([]),
+        }),
+      });
+
+      await expect(
+        bookmarkService.getBookmarkById("49fb6870-478f-4a6a-a839-e6e198adfb94")
+      ).rejects.toThrow("Bookmark not found");
+    });
+
+    it("should throw validation error if the id is not a valid GUID", async () => {
```
Comment (Bala-raj) [inline test/services/bookmarkService.test.js:65]: More Comprehensive Input Validation Tests: You have one general test for invalid input in createBookmark and getBookmarkById.  You should break this down into more specific tests for each validation rule.  For example:

- Test with empty title.
- Test with very long title.
- Test with invalid URLs (missing protocol, malformed domain, etc.).
- Test with special characters in the title (and decide how you want to handle them).

File: test/services/bookmarkService.test.js [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+      db.delete.mockReturnValue({
+        where: vi.fn().mockReturnValue({
+          returning: vi.fn().mockResolvedValue([mockBookmark]),
+        }),
+      });
+
+      const result = await bookmarkService.deleteBookmarkById(
+        "49fb6870-478f-4a6a-a839-e6e198adfb94"
+      );
+      expect(result).toEqual(1);
```
Comment (Bala-raj) [inline test/services/bookmarkService.test.js:141]: Test deleteBookmarkById returning value: The test for deleteBookmarkById currently only checks that the returned value is 1. It would be better to check if the correct number of deleted rows is returned, in case you want to delete multiple bookmarks at once in the future.

**Review by Bala-raj — CHANGES_REQUESTED** [review]

File: test/controllers/bookmarkController.test.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+      await bookmarkController.getAllBookmarks(req, res);
+
+      expect(res.status).toHaveBeenCalledWith(200);
+      expect(res.json).toHaveBeenCalledWith({ data: mockBookmarks });
+    });
+  });
+
+  describe('getBookmarkById', () => {
+    it('should return specific bookmark with status 200', async () => {
+      const mockBookmarks = [{ id: 1, title: 'Test Bookmark', url: 'http://example.com' }];
```
[the remark this answers is not in the document]
Comment (Malankar) [reply test/controllers/bookmarkController.test.js]: It is done

File: test/controllers/bookmarkController.test.js
[the remark this answers is not in the document]
Comment (Malankar) [reply test/controllers/bookmarkController.test.js:1]: this is done as well thanks for recommendation

File: test/controllers/bookmarkController.test.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+vi.mock('../../src/services/bookmarkService');
+
+describe('bookmarkController', () => {
+  describe('getAllBookmarks', () => {
+    it('should return all bookmarks with status 200', async () => {
+      const mockBookmarks = [{ id: 1, title: 'Test Bookmark', url: 'http://example.com' }];
+      bookmarkService.getAllBookmarks.mockResolvedValue(mockBookmarks);
+
+      const req = {};
+      const res = {
```
[the remark this answers is not in the document]
Comment (Malankar) [reply test/controllers/bookmarkController.test.js]: I have abstract res for now as its seems to be same but not the req object

File: test/controllers/bookmarkController.test.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
+import { describe, it, expect, vi } from 'vitest';
+import bookmarkController from '../../src/controllers/bookmarkController';
+import bookmarkService from '../../src/services/bookmarkService';
+
+vi.mock('../../src/services/bookmarkService');
+
+describe('bookmarkController', () => {
+  describe('getAllBookmarks', () => {
+    it('should return all bookmarks with status 200', async () => {
```
[the remark this answers is not in the document]
Comment (Malankar) [reply test/controllers/bookmarkController.test.js]: this is done as well

File: test/middleware/errorHandler.test.js
[the remark this answers is not in the document]
Comment (Malankar) [reply test/middleware/errorHandler.test.js:1]: This is done

File: test/db/schema.test.js
[the remark this answers is not in the document]
Comment (Malankar) [reply test/db/schema.test.js:1]: We don't have it in drizzle but we are doing it in joi validation instead

File: test/services/bookmarkService.test.js (hunk 1 above)
Comment (Malankar) [reply test/services/bookmarkService.test.js]: This is done

File: test/services/bookmarkService.test.js (hunk 2 above)
Comment (Malankar) [reply test/services/bookmarkService.test.js:65]: Hi @Bala-raj 
In getBookmarkById we only accept uuid of the bookmark so I had only one test. I have added more test cases for createBookmark please check

File: test/services/bookmarkService.test.js (hunk 3 above)
Comment (Malankar) [reply test/services/bookmarkService.test.js:142]: Hi @Bala-raj 
deleteBookmarkById functionality is to delete one bookmark only for now there is no functionality to delete multiple for now because of that we will be returning array of one bookmark which is delete that's why it is one. let me know if you want me to implement delete multiple bookmark endpoint

File: test/routes/bookmarkRoutes.test.js
Comment (Malankar) [reply test/routes/bookmarkRoutes.test.js:1]: I have added detailed rate limiting test cases please check
